### PR TITLE
Added the building of nng, safeclib

### PR DESCRIPTION
**Pull request by slyusar003**
Signed-off-by: Ostap <slyusar.ostap@gmail.com>

### CODE REVIEW

File: commands/build.py [hunk 1]
```diff
@@ -31,7 +32,6 @@ def __str__(self):
     def clean(self):
         if os.path.exists(self.build_path):
             logger.info("cleaning {}".format(self.name))
-            os.system("xargs rm < {}/install_manifest.txt".format(self.build_path))
```
Comment (tomereli) [inline commands/build.py:34]: Why was this removed? This cleans installed files, it is usefull

**Comment by tomereli** [comment]
Did you check all maptools build combinations to see they did not break?
```
[tools (HEAD detached at prpl/ostap-dev/add_build_deps) 30dd86a]$ maptools build all -c distclean
CHDLAB_PATH not set, not using chdlab and virtual env...
ASSERT: 'Namespace' object has no attribute 'deps_modules'
Traceback (most recent call last):
  File "/home/tbeliyah/work/prpl/dev1/multiap/tools/maptools.py", line 98, in <module>
    maptools().run()
  File "/home/tbeliyah/work/prpl/dev1/multiap/tools/maptools.py", line 89, in run
    commands[self.args.cmd]()
  File "/home/tbeliyah/work/prpl/dev1/multiap/tools/maptools.py", line 63, in __build__
    mapbuild.build_map(self.args)
  File "/home/tbeliyah/work/prpl/dev1/multiap/tools/commands/build.py", line 119, in build_map
    modules = deps_modules if 'all' in args.deps_modules else [m for m in deps_modules if m in args.modules]
AttributeError: 'Namespace' object has no attribute 'deps_modules'
```

**Comment by slyusar003** [comment]
@tomereli let me try to check all build

**Comment by tomereli** [comment]
Generally, I prefer having a single build command, which accepts multiple options.
- maptools build all - builds everything (safe_c, nng, map repos)
- maptools build map - build all map repos
- maptools build deps - build all deps (safec, nng)
- maptools build <module> - builds a specific module

Each should be tested with -c flags  `-c distclean`, `-c clean`, `-c clean make`, `-c clean prepare`

I would have an abstract builder base class from which cmakebuilder and acbuilder will derive from. Then have a mapping from module name to type and call the base class command which will call the appropriate method in the derived class.

**Review by tomereli — CHANGES_REQUESTED** [review]
Please update commit message (it is empty):
Commit header should be: maptools: build: add safec and nng build support
and should include a summary of what changed and why, and the tests which were done

**Comment by slyusar003** [comment]
> I would have an abstract builder base class from which cmakebuilder and acbuilder will derive from. Then have a mapping from module name to type and call the base class command which will call the appropriate method in the derived class.

@tomereli I think you are right! Will have modification as you mentioned.

**Comment by slyusar003** [comment]
> Generally, I prefer having a single build command, which accepts multiple options.
> 
> * maptools build all - builds everything (safe_c, nng, map repos)
> * maptools build map - build all map repos
> * maptools build deps - build all deps (safec, nng)
> * maptools build  - builds a specific module
> 
> Each should be tested with -c flags `-c distclean`, `-c clean`, `-c clean make`, `-c clean prepare`
> 
> I would have an abstract builder base class from which cmakebuilder and acbuilder will derive from. Then have a mapping from module name to type and call the base class command which will call the appropriate method in the derived class.

@tomereli I think you are right! Will modify as you mentioned.

File: commands/build.py [hunk 2]
```diff
@@ -66,11 +66,57 @@ def make(self):
...
+        self.env["STAGING_DIR"] = ""
+
+    def __str__(self):
+        return "'{}' builder configuration:\n\tsrc_path: {}\n\tbuild_path: {}\n\tinstall_path: {}".format(self.name, self.src_path, self.build_path, self.install_path)
+
+    def clean(self):
+        if os.path.exists(self.src_path):
+            logger.info("cleaning {}".format(self.name))
+            os.chdir(self.src_path)
+            os.system("make clean")
```
Comment (tomereli) [inline commands/build.py:87]: this will clean only the build folder. What about the installed files? we don't want to have to do distclean and delete everything else.
I think that make uninstall should be supported for safec (used it myself).
For nng, you should use the install_manifest.txt (same line you removed for cmakebuilder) - 
`os.system("xargs rm < {}/install_manifest.txt".format(self.build_path))`

File: commands/build.py [hunk 3]
```diff
@@ -66,11 +66,57 @@ def make(self):
...
+        subprocess.check_call(cmd, env=self.env)
+
 class mapbuild(object):
-    def __init__(self, args):
+    @staticmethod
+    def build_map(args):
         if args.verbose: logger.setLevel(logging.DEBUG)
         commands = args.commands
-        modules = build_modules if 'all' in args.modules else [m for m in build_modules if m in args.modules]
+        modules = deps_modules if 'all' in args.deps_modules else [m for m in deps_modules if m in args.modules]
```
Comment (tomereli) [inline commands/build.py:119]: Why does build_map builds deps_modules?

Comment (slyusar003) [reply commands/build.py:119]: yes, that's my fault. I will push the modification.

File: commands/build.py [hunk 4]
```diff
@@ -93,13 +139,57 @@ def __init__(self, args):
...
+    def build_deps(args):
+        if args.verbose: logger.setLevel(logging.DEBUG)
+        commands = args.commands
+        modules = deps_modules if 'all' in args.modules else [m for m in deps_modules if m in args.modules]
+        modules_dir = os.path.join(os.path.realpath(args.map_path), "../3rdparty")
+        build_dir = os.path.realpath(modules_dir + '/../multiap/build')
+        install_dir = os.path.realpath(build_dir + '/install')
+
+        if 'distclean' in commands and os.path.exists(build_dir):
+             logger.info("distclean - deleting {}".format(build_dir))
```
Comment (tomereli) [inline commands/build.py:152]: This will delete everything, including the map build dir. Also, this is duplicating code which exists in build_map. See the previous comment regarding unifying this using inheritance and polymorphism

**Comment by tomereli** [comment]
@slyusar003 - please close this PR, delete the branch, create new branch and PR in https://github.com/prplfoundation/prplMesh

We will soon archive this repository.

**Comment by slyusar003** [comment]
@tomereli I can close PR but am not able to delete this branch.
it seems that I don't have such right on this repo.:)

**Comment by tomereli** [comment]
> @tomereli I can close PR but am not able to delete this branch.
> it seems that I don't have such right on this repo.:)

I changed the permissions - are you done with the branch? If so I will delete it.

**Comment by slyusar003** [comment]
> I changed the permissions - are you done with the branch? If so I will delete it.

@tomereli Can you remove the branch? I still can't do it now.
